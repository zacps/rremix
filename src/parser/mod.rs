use pest_derive::*;

#[derive(Parser)]
#[grammar = "parser/grammar.pest"]
pub struct RemixParser;

#[cfg(test)]
mod tests {
    use crate::format_pair;
    use insta::assert_snapshot;
    use pest::Parser;
    use std::fs;
    use test_case::test_case;

    use super::*;

    #[test_case("../src/standard-lib.rem")]
    //#[test_case("ex/factorial.rem")]
    //#[test_case("ex/factorial2.rem")]
    //#[test_case("ex/primes.rem")]
    #[test_case("ex/test1.rem")]
    #[test_case("ex/test2.rem")]
    #[test_case("ex/test3.rem")]
    #[test_case("ex/test4.rem")]
    #[test_case("ex/test5.rem")]
    #[test_case("ex/test6.rem")]
    #[test_case("ex/test7.rem")]
    #[test_case("ex/test8.rem")]
    #[test_case("ex/test9.rem")]
    #[test_case("ex/test10.rem")]
    #[test_case("ex/test11.rem")]
    #[test_case("ex/test12.rem")]
    #[test_case("ex/test13.rem")]
    #[test_case("ex/test14.rem")]
    #[test_case("ex/test15.rem")]
    #[test_case("ex/test16.rem")]
    #[test_case("ex/test17.rem")]
    #[test_case("ex/test18.rem")]
    #[test_case("ex/test19.rem")]
    #[test_case("ex/test20.rem")]
    #[test_case("ex/test21.rem")]
    #[test_case("ex/test22.rem")]
    #[test_case("ex/test23.rem")]
    #[test_case("ex/test24.rem")]
    #[test_case("ex/test25.rem")]
    #[test_case("ex/test26.rem")]
    // #[test_case("ex/test27.rem")]
    #[test_case("ex/test28.rem")]
    //#[test_case("gr/drawing.rem")]
    //#[test_case("gr/arrows.rem")]
    //#[test_case("gr/squares.rem")]
    //#[test_case("gr/bounce.rem")]
    //#[test_case("ob/obj-demo.rem")]
    //#[test_case("ob/sidewindermaze.rem")]
    //#[test_case("ex/test29.rem")]
    //#[test_case("ex/test30.rem")]
    //#[test_case("ex/test31.rem")]
    //#[test_case("ex/test32.rem")]
    //#[test_case("ex/test33.rem")]
    //#[test_case("ex/test34.rem")]
    //#[test_case("ex/test35.rem")]
    //#[test_case("ex/test36.rem")]
    //#[test_case("ex/test37.rem")]
    //#[test_case("ex/test38.rem")]
    //#[test_case("ex/test39.rem")]
    fn test_parser(path: &'static str) {
        let file =
            fs::read_to_string("tests/".to_owned() + path).expect("could not find test file");
        match RemixParser::parse(Rule::program, &file) {
            Ok(mut parse) => {
                let pair = parse.next();
                assert_snapshot!(path, format_pair(&pair.unwrap(), 0, true))
            }
            Err(e) => {
                eprintln!("{e}");
                panic!("Failed to parse file {path}")
            }
        }
    }
}
