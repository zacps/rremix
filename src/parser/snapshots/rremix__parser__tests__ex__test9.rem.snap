---
source: src/parser/mod.rs
assertion_line: 72
expression: "format_pair(&pair.unwrap(), 0, true)"
---
- program
  - eol_comment: ".Fizz-Bang."
  - sigiled_comment: "==========="
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "for"
    - single_name_part: "each"
    - expression > unary_expression > simple_expression > variable: "n"
    - single_name_part: "from"
    - literal > number: "1"
    - single_name_part: "to"
    - literal > number: "100"
    - newindent_block
      - new_indent: "\t"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "if"
        - expression > unary_expression > simple_expression > function_call
          - expression > unary_expression > simple_expression > variable: "n"
          - single_name_part: "is"
          - single_name_part: "divisible"
          - single_name_part: "by"
          - literal > number: "35"
        - newindent_block
          - new_indent: "\t"
          - statement > expression > unary_expression > simple_expression > function_call
            - literal > string: "\"fizz-bang\""
            - single_name_part: "^"
        - single_name_part: "otherwise"
        - newindent_block
          - new_indent: "\t"
          - statement > expression > unary_expression > simple_expression > function_call
            - single_name_part: "if"
            - expression > unary_expression > simple_expression > function_call
              - expression > unary_expression > simple_expression > variable: "n"
              - single_name_part: "is"
              - single_name_part: "divisible"
              - single_name_part: "by"
              - literal > number: "5"
            - newindent_block
              - new_indent: "\t"
              - statement > expression > unary_expression > simple_expression > function_call
                - literal > string: "\"fizz\""
                - single_name_part: "^"
            - single_name_part: "otherwise"
            - newindent_block
              - new_indent: "\t"
              - statement > expression > unary_expression > simple_expression > function_call
                - single_name_part: "if"
                - expression > unary_expression > simple_expression > function_call
                  - expression > unary_expression > simple_expression > variable: "n"
                  - single_name_part: "is"
                  - single_name_part: "divisible"
                  - single_name_part: "by"
                  - literal > number: "7"
                - newindent_block
                  - new_indent: "\t"
                  - statement > expression > unary_expression > simple_expression > function_call
                    - literal > string: "\"bang\""
                    - single_name_part: "^"
                - single_name_part: "otherwise"
                - newindent_block
                  - new_indent: "\t"
                  - statement > expression > unary_expression > simple_expression > function_call
                    - expression > unary_expression > simple_expression > variable: "n"
                    - single_name_part: "^"
      - statement > expression > unary_expression > simple_expression > function_call
        - literal > string: "\" \""
        - single_name_part: "^"
  - statement > expression > unary_expression > simple_expression > variable: "↲"
  - eol_comment: "Trying to do same thing with less indentation."
  - statement > expression > unary_expression > simple_expression > function_call
    - single_name_part: "for"
    - single_name_part: "each"
    - expression > unary_expression > simple_expression > variable: "n"
    - single_name_part: "from"
    - literal > number: "1"
    - single_name_part: "to"
    - literal > number: "100"
    - newindent_block
      - new_indent: "\t"
      - statement > expression > unary_expression > simple_expression > function_call
        - single_name_part: "choose"
        - single_name_part: "from"
        - newindent_block
          - new_indent: "\t"
          - statement > expression > unary_expression > simple_expression > function_call
            - expression > unary_expression > simple_expression > function_call
              - expression > unary_expression > simple_expression > variable: "n"
              - single_name_part: "is"
              - single_name_part: "divisible"
              - single_name_part: "by"
              - literal > number: "35"
            - single_name_part: "do"
            - block > statement > expression > unary_expression > simple_expression > function_call
              - literal > string: "\"fizz-bang\""
              - single_name_part: "^"
          - statement > expression > unary_expression > simple_expression > function_call
            - expression > unary_expression > simple_expression > function_call
              - expression > unary_expression > simple_expression > variable: "n"
              - single_name_part: "is"
              - single_name_part: "divisible"
              - single_name_part: "by"
              - literal > number: "5"
            - single_name_part: "do"
            - block > statement > expression > unary_expression > simple_expression > function_call
              - literal > string: "\"fizz\""
              - single_name_part: "^"
          - statement > expression > unary_expression > simple_expression > function_call
            - expression > unary_expression > simple_expression > function_call
              - expression > unary_expression > simple_expression > variable: "n"
              - single_name_part: "is"
              - single_name_part: "divisible"
              - single_name_part: "by"
              - literal > number: "7"
            - single_name_part: "do"
            - block > statement > expression > unary_expression > simple_expression > function_call
              - literal > string: "\"bang\""
              - single_name_part: "^"
          - statement > expression > unary_expression > simple_expression > function_call
            - expression > unary_expression > simple_expression > variable: "n"
            - single_name_part: "^"
      - statement > expression > unary_expression > simple_expression > function_call
        - literal > string: "\" \""
        - single_name_part: "^"
  - statement > expression > unary_expression > simple_expression > variable: "↲"
  - EOI: ""
